LUSDChickenBonds/src/NFTArtwork/ChickenInTraitWeights.sol

```diff
@@ -0,0 +1,130 @@
+                chickenWeightsCached[i] = chickenWeightsCached[i] * EGG_CHICKEN_REPEAT_FACTOR / 1e18;
+            }
+            aaa -= chickenWeightsCached[i];
+            bbb -= shellWeights[i];
+        }
+        // Regular traits
+        for (uint256 i = 0; i < uint256(ShellColor.Bronze); i++) {
+            chickenWeightsCached[i] = shellWeights[i] * aaa / bbb;
+        }
+    }
+
+    // Turn the pseudo-random number `rand` -- 18 digit FP in range [0,1) -- into a Chicken (Shell) color.
+    function _getChickenColor(uint256 rand, ShellColor shellColor, uint256 troveFactor) internal view returns (ShellColor) {
+        // first adjust weights for affinity
+        uint256[13] memory chickenWeightsCached = _getChickenAffinityWeights(shellColor, troveFactor);
+
+        // then compute color
+        uint256 needle = chickenWeightsCached[uint256(ShellColor.OffWhite)];
+        if (rand < needle) { return ShellColor.OffWhite; }
+        needle += chickenWeightsCached[uint256(ShellColor.LightBlue)];
+        if (rand < needle) { return ShellColor.LightBlue; }
+        needle += chickenWeightsCached[uint256(ShellColor.DarkerBlue)];
+        if (rand < needle) { return ShellColor.DarkerBlue; }
+        needle += chickenWeightsCached[uint256(ShellColor.LighterOrange)];
+        if (rand < needle) { return ShellColor.LighterOrange; }
+        needle += chickenWeightsCached[uint256(ShellColor.LightOrange)];
+        if (rand < needle) { return ShellColor.LightOrange; }
+        needle += chickenWeightsCached[uint256(ShellColor.DarkerOrange)];
+        if (rand < needle) { return ShellColor.DarkerOrange; }
+        needle += chickenWeightsCached[uint256(ShellColor.LightGreen)];
+        if (rand < needle) { return ShellColor.LightGreen; }
+        needle += chickenWeightsCached[uint256(ShellColor.DarkerGreen)];
+        if (rand < needle) { return ShellColor.DarkerGreen; }
+        needle += chickenWeightsCached[uint256(ShellColor.Bronze)];
+        if (rand < needle) { return ShellColor.Bronze; }
+        needle += chickenWeightsCached[uint256(ShellColor.Silver)];
+        if (rand < needle) { return ShellColor.Silver; }
+        needle += chickenWeightsCached[uint256(ShellColor.Gold)];
+        if (rand < needle) { return ShellColor.Gold; }
+        needle += chickenWeightsCached[uint256(ShellColor.Rainbow)];
+        if (rand < needle) { return ShellColor.Rainbow; }
+        return ShellColor.Luminous;
+    }
+
+    function _getChickenTrait9(uint256 rand, uint256 troveFactor) internal view returns (uint8) {
+        // TODO: should we hardcode this? Gas doesn’t really matter here as we are just static-calling them
+        uint256 tier1Probability = ((2e18 - ALPHA3) / 3 - R3 * troveFactor / 1e18) / 3;
+        uint256 tier2Probability = 1e18 / 9;
+        uint256 tier3Probability = (ALPHA3 / 3 + R3 * troveFactor / 1e18) / 3;
+
+        uint256 needle = tier1Probability;
+        if (rand < needle) { return 1; }
```

Comment:
Great that we're using 1-based indices! Makes it simpler to convert it to metadata.